src/dbnode/encoding/types.go

```diff
@@ -27,20 +27,33 @@ import (
 	"github.com/m3db/m3/src/dbnode/ts"
 	"github.com/m3db/m3/src/dbnode/x/xio"
 	"github.com/m3db/m3/src/dbnode/x/xpool"
-	"github.com/m3db/m3/src/x/serialize"
 	"github.com/m3db/m3/src/x/checked"
 	"github.com/m3db/m3/src/x/ident"
 	"github.com/m3db/m3/src/x/pool"
+	"github.com/m3db/m3/src/x/serialize"
 	xtime "github.com/m3db/m3/src/x/time"
 )
 
+// StreamOpts is an options struct that can be passed to Encoder.Stream()
+// to modify its behavior.
+type StreamOpts struct {
+	// Optional bytes into which the stream should be copied. If left
+	// nil then the checked bytes pool on the encoder will be used instead.
+	// TODO(rartoul): Actually use this field in the Stream API() in a subsequent
+	// P.R.
```

Comment:
Just a thought while here, we should just be able to pass `StreamOptions` up and down the call stack to Snapshot, Tick, etc yeah?  Obviously with parallel ticking we'll have to use a few and reuse them, but for Snapshot that should just be linear and could use a single one I would guess?

Reply:
Yeah flush and snapshot can use the same one passed up and down the stack.

Ticking is more complicated because you’re not copying temporarily, you iterate multiple streams in parallel and encode them into a new long-lives stream

Reply:
Hm yeah.

Reply:
@robskillington Lol yeah we could do something tricky like that. I think what we really need is a utility method for performing merges that we can bake some of that crazier logic into. Another issue we have with merges is that basically if you're merging two large streams into one new stream there is no guessing regarding how big the new stream will be since you just write into an encoder one at a time.

If you had a utility we could use over the code base we could use the size of all the streams that are being merged to guess the size of the byte slice we should use to merge into so we don't have to allocate all the smaller slices along the way if that makes sense

Reply:
Agreed.  I do think that it's not so much worth worrying about how big they will be (and guest-imating that), but just reusing buffers that grow as necessary is ideal.  That way if it grows sure it needs to allocate, but if its reused across all series during a tick/merge then there's still zero allocations per series after amortizing cost.  

i.e. any time we need to do an allocation per merge series, we have a problem, as long as we reuse buffers, the allocationsPerMergeSeries approaches zero over any significant number of series.

Reply:
Yeah sure allocations will go down but if we’re doing 5-6 grows per series (which is not unlikely and will be even higher with complex types workloads) we can make the merges faster because that’s 6 pool gets AND puts we don’t have to do + the actual time spent copying that we don’t have to do.

A micro optimization for sure, but wouldn’t be surprised if it made merging much faster